server/store_statistics.go

```diff
@@ -0,0 +1,131 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"strconv"
+
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"github.com/pingcap/pd/server/core"
+	"github.com/pingcap/pd/server/namespace"
+)
+
+type storeStatistics struct {
+	opt             *scheduleOption
+	namespace       string
+	Up              int
+	Disconnect      int
+	Down            int
+	Offline         int
+	Tombstone       int
+	LowSpace        int
+	StorageSize     uint64
+	StorageCapacity uint64
+	RegionCount     int
+	LeaderCount     int
+}
+
+func newStoreStatistics(opt *scheduleOption, namespace string) *storeStatistics {
+	return &storeStatistics{
+		opt:       opt,
+		namespace: namespace,
+	}
+}
+
+func (s *storeStatistics) Observe(store *core.StoreInfo) {
+	// Store state.
+	switch store.GetState() {
+	case metapb.StoreState_Up:
+		if store.DownTime() >= s.opt.GetMaxStoreDownTime() {
+			s.Down++
+		} else if store.IsDisconnected() {
+			s.Disconnect++
+		} else {
+			s.Up++
+		}
+	case metapb.StoreState_Offline:
+		s.Offline++
+	case metapb.StoreState_Tombstone:
+		s.Tombstone++
+		return
+	}
+	if store.IsLowSpace() {
+		s.LowSpace++
+	}
+
+	// Store stats.
+	s.StorageSize += store.StorageSize()
+	s.StorageCapacity += store.Stats.GetCapacity()
+	s.RegionCount += store.RegionCount
+	s.LeaderCount += store.LeaderCount
+
+	id := strconv.FormatUint(store.GetId(), 10)
+	balanceScoreGauge.WithLabelValues(s.namespace, id, "region").Set(store.RegionScore())
+	balanceScoreGauge.WithLabelValues(s.namespace, id, "leader").Set(store.LeaderScore())
+}
+
+func (s *storeStatistics) Collect() {
+	metrics := make(map[string]float64)
+	metrics["store_up_count"] = float64(s.Up)
+	metrics["store_disconnected_count"] = float64(s.Disconnect)
+	metrics["store_down_count"] = float64(s.Down)
+	metrics["store_offline_count"] = float64(s.Offline)
+	metrics["store_tombstone_count"] = float64(s.Tombstone)
+	metrics["store_low_space_count"] = float64(s.LowSpace)
+	metrics["region_count"] = float64(s.RegionCount)
+	metrics["leader_count"] = float64(s.LeaderCount)
+	metrics["storage_size"] = float64(s.StorageSize)
+	metrics["storage_capacity"] = float64(s.StorageCapacity)
+
+	for label, value := range metrics {
+		clusterStatusGauge.WithLabelValues(label, s.namespace).Set(value)
+	}
+}
+
+// It is used for counting up stats of all namespaces.
+const allNamespace = "_all"
```

Comment:
seem we can aggregate all metrics in Prometheus directly.

Reply:
I want to use templating to select namespace, but the templating query supports very limited queries. I found no approach to join a constant with a query result vector. /cc @huachaohuang

Reply:
If we can't do it in Grafana, maybe we can create a new board for all.